### PR TITLE
Properly stop logger during (re)connect failure

### PR DESCRIPTION
**Pull request by akerouanton**
PR #77 introduced a new parameter named ForceStopAsyncSend. It can be
used to tell the logger to not try to send all the log messages in its
buffer before closing. Without this parameter, the logger hangs out
whenever it has logs to write and the target Fluentd server is down.

But this new parameter is not enough: the connection is currently lazily
initialized when the logger receive its first message. This blocks the
select reading messages from the queue (until the connection is ready).
Moreover, the connection dialing uses an exponential back-off retry.

Because of that, the logger won't look for messages on `stopRunning`
channel (the channel introduced by PR #77), either because it's
blocked by the Sleep used for the retry or because the connection
dialing is waiting until dialing timeout (eg. TCP timeout).

To fix these edge cases, the time.Sleep() used for back-off retry has
been transformed into a time.After() and a new stopAsyncConnect channel
has been introduced to stop it. Moreover, the dialer.Dial() call used
to connect to Fluentd has been replaced with dialer.DialContext() and a
cancellable context is now used to stop the call to that method.

Finally, the Fluent.run() method has been adapted to wait for both new
messages on f.pending and stop signal sent to f.stopRunning channel.
Previously, both channels were selected independently, in a procedural
fashion.

This fix is motivated by the issue described in: moby/moby#40063.

### CODE REVIEW

**Comment by tagomoris** [comment]
The motivation is reasonable, so I'm positive to fix that problem.
This kind of patch requires time to review, so I'll be able to review this change tomorrow or early next week.
But anyway, CI is now failing. @NiR- Could you take care of it?

**Comment by tagomoris** [comment]
@NiR- Great! Please let me know when you finished that.

**Comment by angulito** [comment]
Any news on this? I'm experiencing the same issue with `docker v19.03.6-ce` that is happening https://github.com/moby/moby/issues/40063. Thank you!

**Comment by akerouanton** [comment]
@angulito We first want to merge #83 which is introducing a new way of testing edge cases like the one identified and fixed by this PR.

The new test cases and the fix were first submitted in a single PR, but since the diff was quite large and important, I splitted it in two PRs to make it less difficult for the repo's maintainer(s) to review and accept these changes.

Hopefully this will be included in the next major docker release. In the mean time, I created a logger plugin including this fix. It's working like the original fluentd logger and it's available in [akerouanton/fluentd-async-logger](https://github.com/akerouanton/fluentd-async-logger).

**Comment by angulito** [comment]
Thank you for working hard on it! @akerouanton

**Comment by akerouanton** [comment]
Hey @tagomoris, I took a look at this PR in the past days and I finally found out why tests were failing on the CI. I also overhauled the previous implementation to make the code substantially simpler.

CI was failing because of tests with `RequestAck`  option enabled: the ack hashes contain a randomly generated uint and despite the generator being implicitly seeded with 1 (as specified [here](https://golang.org/pkg/math/rand/#Seed)), the ack hashes were not stable. I tried to explicitly seed the generator but with no luck. Thus, I decided to introduce a new `randomGenerator` variable and initialize it with `rand.Uint64`. This variable is changed in tests to use a function that always return 1. This way, we're sure the ack hashes are stable across time and machines.

On the overhaul part, I decided to remove the newly introduced `stopAsyncConnect` channel and use a cancellable context instead. This context is passed from `newWithDialer()` to `run() -> write() -> connectOrRetry() -> connect() -> DialWithContext()`. In this way, as soon as the context got cancelled, any in-progress dialing got stopped and the `errIsClosing` error is propagated up to `run()` by early terminating retriable code paths (`connectOrRetry()` and `write()`) . Moreover, in this way, any subsequent calls to `write()` will fail the same way (for instance if a log is dispatched through `f.pending` before `f.stopRunning` closing is dispatched), making the code more robust to channel dispatching order in `run()`. I also removed the `drainEvents` condition `run()` since returning from that function and closing `f.pending` is already enough to discard queued messages.

**Comment by tagomoris** [comment]
I'll try to find time to review this change this or the next week.

**Comment by tagomoris** [comment]
I took a glance at this, and need more time.

**Comment by tagomoris** [comment]
I reviewed the updated change. The clear problem is, this code will acquire the lock of connection during writing messages and waiting for the ack message. This causes terrible performance degradation if the ack option is enabled. Even without enabling the ack option, the critical section for writing messages could cause performance problems.

And I found another problem (this is not introduced by this change) that arriving ack messages may be disordered, but the current implementation doesn't consider such cases. 🤔

**Comment by jchanam** [comment]
Hi!

I'd like to know if you were able to move forward with this issue. We're still facing this problem described in https://github.com/moby/moby/issues/40063

Thanks!

**Comment by rothn** [comment]
We are having this issue as well. Do we know if there's anything preventing a merge?

**Comment by hakashya** [comment]
Hey @tagomoris,
To confirm, the problems that are preventing the merge are:
1. Message writing critical section causing performance degradation.
2. Unordered ack messages not being handled.
<br/>

Hey @akerouanton,
I am just wondering if you were working on it?

**Comment by tagomoris** [comment]
@hakashya Unordered ack messages are not the problem of this change. Of course, it's good if we can solve it.

**Comment by PettitWesley** [comment]
@akerouanton @tagomoris Hi, I'm Wesley, from the AWS containers team, and I'm also the AWS maintainer of fluent/fluent-bit. We need to get this change out since it fixes the linked Moby/Docker issue.

My proposal is that I will take over this fix. I will preserve @akerouanton's commit so that they get credit. And then add commits on top of it as needed to get this PR to a merge-able state.

To confirm, the issues blocking merge are:
1. Rebase.
2. "Message writing critical section causing performance degradation." I need to read through this code and try to understand it... can someone point me to the piece of the code where this issue is?

If I fix those, we can merge it?

**Comment by tagomoris** [comment]
I'm totally ok to merge the change written by 2 contributors (it's almost equal to independent 2 pull-reqs) if it can solve problems.

**Comment by tagomoris** [comment]
The performance problem is in [the updated `write` function](https://github.com/fluent/fluent-logger-golang/pull/82/files#diff-0c05db8bc36bb5fe014b3a8a32666b6d253dc7382bc4ccdab64bfe50d92b5c92R462).

The `f.muconn` is locked at [the beginning of the `writer`](https://github.com/fluent/fluent-logger-golang/pull/82/files#diff-0c05db8bc36bb5fe014b3a8a32666b6d253dc7382bc4ccdab64bfe50d92b5c92R467-R468) and it'll be unlocked when the control will return from this function.
But in the `writer`, it'll [wait for the Ack message from the destination in sync manner](https://github.com/fluent/fluent-logger-golang/pull/82/files#diff-0c05db8bc36bb5fe014b3a8a32666b6d253dc7382bc4ccdab64bfe50d92b5c92R490-R505) - and it may be delayed, or may not arrive forever. While waiting for the Ack response, the connection `f.muconn` is locked and be unavailable from other goroutines (threads).

**Comment by akerouanton** [comment]
@PettitWesley are you working on this? I have some time this week to try to fix the issues pointed by @tagomoris.

**Comment by PettitWesley** [comment]
@akerouanton other things came onto my backlog which have delayed me; please work on getting this PR ready for merge this week if you can.

**Comment by PettitWesley** [comment]
@akerouanton Thanks for following up on this! 💙

I know many users who will be very happy once this fixed is merged, and released upstream in Docker :)

File: fluent/fluent.go [hunk 1]
```diff
@@ -362,49 +397,95 @@ func (f *Fluent) appendBuffer(msg *msgToSend) error {
...
 
-func (f *Fluent) run() {
-	drainEvents := false
-	var emitEventDrainMsg sync.Once
+var errIsClosing = errors.New("fluent logger is closing")
+
+// Caller should take care of locking muconn first.
+func (f *Fluent) connectOrRetry(ctx context.Context) error {
+	// Use a Time channel instead of time.Sleep() to avoid blocking this
+	// goroutine during possibly way too much time (because of the exponential
```
[the remark this answers is not in the document]
Comment (akerouanton) [reply fluent/fluent.go:432]: Mh "waste too much time" would not fit with the beginning of the phrase. Do you want to rephrase the whole comment?

[the remark this answers is not in the document]
Comment (tagomoris) [reply fluent/fluent.go]: Ah, no, it's fine. I misunderstood this comment.

**Comment by tagomoris** [comment]
LGTM!
Let me merge this change and tag `v1.7.0` to let users choose this fix only (or `v1.6.3` for other recent changes).

**Comment by tagomoris** [comment]
Done. Thank you for the great work!
